File: bots/tracker-wallet-python/rust/src/tracker.rs
use std::collections::HashMap;
use std::sync::Arc;

use chrono::Utc;
use futures::TryStreamExt;
use mongodb::bson::doc;
use mongodb::Database;
use poise::serenity_prelude as serenity;
use serenity::builder::{CreateEmbed, CreateEmbedFooter, CreateMessage};
use serenity::http::Http;
use serenity::model::id::ChannelId;
use serenity::model::Timestamp;
use tokio::time::{self, Duration};
use tracing::{debug, error, info, warn};

use crate::models::{Subscription, Wallet};
use crate::polymarket::{fetch_closed_position, fetch_user_activity, PolyActivity};

const CHECK_INTERVAL_SECS: u64 = 15;
const SENT_CACHE_TTL_SECS: i64 = 120;

// ─── Entry Tier ──────────────────────────────────────────────────────────────

const ENTRY_HIGH_USD: f64 = 1000.0;
const ENTRY_MEDIUM_USD: f64 = 500.0;
const ENTRY_LOW_USD: f64 = 10.0;

struct EntryTier {
    label: &'static str,
    emoji: &'static str,
}

fn get_entry_tier(value_usd: f64) -> EntryTier {
    if value_usd >= ENTRY_HIGH_USD {
        EntryTier {
            label: "ALTA",
            emoji: "🚀",
        }
    } else if value_usd >= ENTRY_MEDIUM_USD {
        EntryTier {
            label: "MEDIANA",
            emoji: "📈",
        }
    } else if value_usd < ENTRY_LOW_USD {
        EntryTier {
            label: "BAIXA",
            emoji: "🧊",
        }
    } else {
        EntryTier {
            label: "MÉDIA",
            emoji: "➖",
        }
    }
}

// ─── Loop principal ───────────────────────────────────────────────────────────

pub async fn start_tracker_loop(http: Arc<Http>, db: Database, http_client: reqwest::Client) {
    info!("🔥 TRACKER INICIADO");
    info!("⏱️  Intervalo: {}s", CHECK_INTERVAL_SECS);

    let mut interval = time::interval(Duration::from_secs(CHECK_INTERVAL_SECS));

    // Cache de dedup: id do trade → timestamp (evita mensagens duplicadas)
    let mut sent_messages: HashMap<String, i64> = HashMap::new();

    loop {
        interval.tick().await;

        let now = Utc::now().timestamp();

        // Limpa entradas antigas do cache
        sent_messages.retain(|_, ts| now - *ts < SENT_CACHE_TTL_SECS);

        // Carrega todas as carteiras monitoradas
        let wallets_col = db.collection::<Wallet>("wallets");
        let wallets: Vec<Wallet> = match wallets_col.find(doc! {}).await {
            Ok(cursor) => match cursor.try_collect().await {
                Ok(v) => v,
                Err(e) => {
                    error!("❌ Erro ao coletar wallets: {e}");
                    continue;
                }
            },
            Err(e) => {
                error!("❌ Erro ao buscar wallets: {e}");
                continue;
            }
        };

        debug!("📊 Carteiras monitoradas: {}", wallets.len());

        if wallets.is_empty() {
            debug!("⚠️ Nenhuma carteira cadastrada");
            continue;
        }

        for wallet in wallets {
            if !wallet.address.starts_with("0x") {
                warn!("⚠️ Endereço inválido: {}", wallet.address);
                continue;
            }

            // Verifica se há inscrições ativas para esta carteira
            let subs_col = db.collection::<Subscription>("subscriptions");
            let subs: Vec<Subscription> = match subs_col
                .find(doc! { "walletAddress": &wallet.address })
                .await
            {
                Ok(cursor) => match cursor.try_collect().await {
                    Ok(v) => v,
                    Err(e) => {
                        error!("❌ Erro ao coletar subscriptions: {e}");
                        continue;
                    }
                },
                Err(e) => {
                    error!("❌ Erro ao buscar subscriptions: {e}");
                    continue;
                }
            };

            if subs.is_empty() {
                debug!("⚠️ Carteira {} sem inscrições ativas", &wallet.address[..8]);
                continue;
            }

            debug!(
                "🔍 Checando {}... ({} canal(is))",
                &wallet.address[..8],
                subs.len()
            );

            // Busca atividades desde o último timestamp via API oficial
            let activities =
                fetch_user_activity(&http_client, &wallet.address, wallet.last_timestamp).await;

            if activities.is_empty() {
                continue;
            }

            info!(
                "🚨 MUDANÇA DETECTADA: {} operação(ões) para {}",
                activities.len(),
                &wallet.address[..8]
            );

            // Atualiza lastTimestamp para o maior timestamp recebido
            let max_ts = activities.iter().map(|a| a.timestamp).max().unwrap_or(now);
            if let Err(e) = wallets_col
                .update_one(
                    doc! { "address": &wallet.address },
                    doc! { "$set": { "lastTimestamp": max_ts } },
                )
                .await
            {
                error!("❌ Erro ao atualizar lastTimestamp: {e}");
            }

            // Busca username uma vez por carteira via campo da API
            // (display_name já vem em cada trade — usado no build_embed)

            for trade in &activities {
                // Dedup por transaction hash
                if sent_messages.contains_key(&trade.id) {
                    debug!(
                        "⏭️ Pulando duplicata: {}",
                        &trade.id[..20.min(trade.id.len())]
                    );
                    continue;
                }

                // Para SELL: busca PNL via /closed-positions
                let realized_pnl = if trade.side == "SELL" && !trade.condition_id.is_empty() {
                    fetch_closed_position(&http_client, &wallet.address, &trade.condition_id).await
                } else {
                    None
                };

                let embed = build_embed(trade, &wallet.address, realized_pnl);

                let mut sent_count: u32 = 0;

                for sub in &subs {
                    // ─── Filtros por subscription ─────────────────────────
                    if let Some(ref filters) = sub.filters {
                        // Filtro de valor mínimo (USD)
                        if let Some(min_usd) = filters.min_usd {
                            if min_usd > 0.0 && trade.usdc_size < min_usd {
                                continue;
                            }
                        }
                        // Filtro de palavras-chave no título do mercado
                        if let Some(ref keywords) = filters.keywords {
                            if !keywords.is_empty() {
                                let title_lower = trade.market_title.to_lowercase();
                                if !keywords
                                    .iter()
                                    .any(|k| title_lower.contains(&k.to_lowercase()))
                                {
                                    continue;
                                }
                            }
                        }
                    }
                    // ──────────────────────────────────────────────────────

                    let channel_id_u64: u64 = match sub.channel_id.parse() {
                        Ok(v) => v,
                        Err(_) => {
                            error!("❌ channelId inválido: {}", sub.channel_id);
                            continue;
                        }
                    };

                    let channel = ChannelId::new(channel_id_u64);

                    match channel
                        .send_message(&http, CreateMessage::new().add_embed(embed.clone()))
                        .await
                    {
                        Ok(_) => {
                            sent_count += 1;
                            info!("✓ Enviado para canal {}", sub.channel_id);
                        }
                        Err(e) => {
                            error!("❌ Erro ao enviar para {}: {e}", sub.channel_id);
                        }
                    }
                }

                if sent_count > 0 {
                    sent_messages.insert(trade.id.clone(), now);
                    info!("✅ Mensagem enviada para {sent_count} canal(is)");
                } else {
                    warn!("⚠️ Nenhum canal recebeu a mensagem (filtros ou erros)");
                }

                // Delay entre mensagens para evitar rate limit do Discord
                tokio::time::sleep(Duration::from_millis(500)).await;
            }

            // Pausa entre carteiras para evitar rate limit da Polymarket API
            tokio::time::sleep(Duration::from_secs(1)).await;
        }

        debug!("✓ Ciclo de verificação concluído");
    }
}

// ─── Construção do embed Discord ──────────────────────────────────────────────

fn build_embed(
    trade: &PolyActivity,
    wallet_address: &str,
    realized_pnl: Option<f64>,
) -> CreateEmbed {
    let (type_label, color, emoji) = match trade.side.as_str() {
        "BUY" => ("COMPROU", 0x00ff00u32, "🟢"),
        "SELL" => ("VENDEU", 0xff0000u32, "🔴"),
        _ => ("OPERAÇÃO", 0x808080u32, "📊"),
    };

    let market_url = if trade.market_url.is_empty() {
        format!("https://polymarket.com/profile/{wallet_address}")
    } else {
        trade.market_url.clone()
    };

    let market_title_linked = if trade.market_url.is_empty() {
        trade.market_title.clone()
    } else {
        format!("[{}]({})", trade.market_title, trade.market_url)
    };

    let addr_short = format!(
        "{}...{}",
        &wallet_address[..6],
        &wallet_address[wallet_address.len() - 4..]
    );

    let mut description = String::new();
    // Usa display_name vindo direto da API (é o pseudonym/name do trader)
    if let Some(ref name) = trade.display_name {
        description.push_str(&format!("**Trader:** @{name}\n"));
    }
    description.push_str(&format!("**Mercado:** {market_title_linked}\n"));
    description.push_str(&format!("**Posição:** {}\n", trade.outcome));
    description.push_str(&format!(
        "**Carteira:** [`{addr_short}`](https://polymarket.com/profile/{wallet_address})"
    ));

    let tier = get_entry_tier(trade.usdc_size);

    let mut embed = CreateEmbed::new()
        .title(format!("{emoji} {type_label}"))
        .url(&market_url)
        .color(color)
        .description(&description)
        .field(
            format!("{} Entrada", tier.emoji),
            format!("{} (${:.2})", tier.label, trade.usdc_size),
            true,
        )
        .field("💵 Preço", format!("${:.3}", trade.price), true)
        .field("📊 Shares", format!("{:.1}", trade.shares), true)
        .field("💰 Valor", format!("${:.2}", trade.usdc_size), true)
        .footer(CreateEmbedFooter::new("Polymarket Tracker"));

    // PNL: exibe somente em SELLs com lucro positivo
    if trade.side == "SELL" {
        if let Some(pnl) = realized_pnl {
            if pnl > 0.0 {
                embed = embed.field("🏆 Profit", format!("${pnl:.2}"), true);
            } else if pnl < 0.0 {
                embed = embed.field("📉 Loss", format!("-${:.2}", pnl.abs()), true);
            }
        }
    }

    if let Ok(ts) = Timestamp::from_unix_timestamp(trade.timestamp) {
        embed = embed.timestamp(ts);
    }

    if !trade.market_image_url.is_empty() {
        embed = embed.image(&trade.market_image_url);
    }

    embed
}

File: bots/tracker-wallet-python/src/tracker.ts
import { Client, EmbedBuilder } from "discord.js";
import { Wallet, Subscription } from "./models.js";
import { fetchRecentActivity, getUsernameFromAddress } from "./polymarket.js";
import { logger } from "./logger.js";

const CHECK_INTERVAL = 10000; // 10 segundos

const ENTRY_HIGH_USD = 1000;
const ENTRY_MEDIUM_USD = 500;
const ENTRY_LOW_USD = 10;

function getEntryTier(valueUsd: number): {
  label: string;
  emoji: string;
} {
  if (valueUsd >= ENTRY_HIGH_USD) {
    return { label: "ALTA", emoji: "🚀" };
  }

  if (valueUsd >= ENTRY_MEDIUM_USD) {
    return { label: "MEDIANA", emoji: "📈" };
  }

  if (valueUsd < ENTRY_LOW_USD) {
    return { label: "BAIXA", emoji: "🧊" };
  }

  return { label: "MÉDIA", emoji: "➖" };
}

// Cache para evitar duplicação de mensagens
const sentMessages = new Map<string, number>();
const MESSAGE_CACHE_TTL = 120000; // 2 minutos

export async function startTrackerLoop(client: Client) {
  logger.info(`🔥 TRACKER V4 INICIADO`);
  logger.info(`🎯 Detecta: Novas posições, aumentos, diminuições e fechamentos`);
  logger.info(`⏱️  Intervalo: ${CHECK_INTERVAL / 1000}s\n`);

  // Aguarda o bot estar pronto
  if (!client.isReady()) {
    logger.info(`⏳ Aguardando bot ficar online...`);
    await new Promise((resolve) => {
      client.once("clientReady", resolve);
    });
    logger.info(`✅ Bot online! Iniciando monitoramento...\n`);
  }

  setInterval(async () => {
    try {
      const timestamp = new Date().toLocaleTimeString("pt-BR");
      logger.debug(`\n💓 [${timestamp}] Verificando carteiras...`);

      // Limpa cache de mensagens antigas
      const now = Date.now();
      for (const [id, timestamp] of sentMessages.entries()) {
        if (now - timestamp > MESSAGE_CACHE_TTL) {
          sentMessages.delete(id);
        }
      }

      const wallets = await Wallet.find();
      logger.debug(`📊 Total de carteiras monitoradas: ${wallets.length}`);

      if (wallets.length === 0) {
        logger.debug(`⚠️ Nenhuma carteira cadastrada`);
        return;
      }

      for (const wallet of wallets) {
        if (!wallet.address.startsWith("0x")) {
          logger.warn(`⚠️ Endereço inválido: ${wallet.address}`);
          continue;
        }

        // Verifica se tem inscrições ativas
        const subs = await Subscription.find({ walletAddress: wallet.address });
        if (subs.length === 0) {
          logger.debug(`⚠️ Carteira ${wallet.address.slice(0, 8)} sem inscrições ativas`);
          continue;
        }

        logger.debug(`🔍 Checando ${wallet.address.slice(0, 8)}... (${subs.length} canal(is))`);

        // Busca atividades via /activity endpoint (desde o último timestamp)
        const activities = await fetchRecentActivity(wallet.address, wallet.lastTimestamp);
        const trackedDisplayName = await getUsernameFromAddress(wallet.address);

        if (activities.length === 0) {
          // Não loga mais nada aqui, o fetchRecentActivity já loga
          continue;
        }

        logger.info(`🚨 MUDANÇA DETECTADA: ${activities.length} operação(ões)\n`);

        // Atualiza timestamp com o maior timestamp recebido (converte s → ms para compatibilidade)
        const maxTs = Math.max(...activities.map((a) => a.timestamp)) * 1000;
        await Wallet.updateOne({ _id: wallet._id }, { lastTimestamp: maxTs });
        // Processa cada trade detectado
        for (const trade of activities) {
          // Verifica se já enviou
          if (sentMessages.has(trade.id)) {
            logger.debug(`   ⏭️ Pulando duplicata: ${trade.id.slice(0, 20)}...`);
            continue;
          }

          logger.debug(`\n   📤 Preparando mensagem...`);
          logger.debug(`      Tipo: ${trade.side}`);
          logger.debug(`      Mercado: ${trade.marketTitle.slice(0, 60)}`);
          logger.debug(`      Outcome: ${trade.outcome}`);
          logger.debug(`      Shares: ${trade.amount.toFixed(1)}`);
          logger.debug(`      Preço: ${trade.price.toFixed(3)}`);
          logger.debug(`      EventSlug: ${trade.eventSlug || "(sem eventSlug)"}`);
          logger.debug(`      MarketSlug: ${trade.marketSlug || "(sem marketSlug)"}`);

          // Detecção de tipo e cor
          let typeLabel = "OPERAÇÃO";
          let color = 0x808080;
          let emoji = "📊";

          const side = (trade.side || "").toUpperCase();

          if (side === "BUY") {
            typeLabel = "COMPROU";
            color = 0x00ff00;
            emoji = "🟢";
          } else if (side === "SELL") {
            typeLabel = "VENDEU";
            color = 0xff0000;
            emoji = "🔴";
          }

          // URL do mercado: pré-construída pela /activity a partir dos slugs
          const marketUrl = trade.marketUrl || `https://polymarket.com/profile/${wallet.address}`;
          const marketTitle = trade.marketUrl ? `[${trade.marketTitle}](${trade.marketUrl})` : trade.marketTitle;

          // Monta descrição
          let description = "";

          // Menciona o usuário que criou o tracking (se disponível)
          const trackingUserIds = new Set<string>();
          for (const sub of subs) {
            if (sub.userId) {
              trackingUserIds.add(sub.userId);
            }
          }

          // Adiciona menções dos usuários que criaram o tracking
          // if (trackingUserIds.size > 0) {
          //   const mentions = Array.from(trackingUserIds)
          //     .map((uid) => `<@${uid}>`)
          //     .join(" ");
          //   description += `${mentions}\n\n`;
          // }

          const traderName = trackedDisplayName || trade.displayName;
          if (traderName) {
            description += `**Trader:** ${traderName}\n`;
          }

          // Usa o marketTitle que já tem o link (ou não)
          description += `**Mercado:** ${marketTitle}\n`;
          description += `**Posição:** ${trade.outcome}\n`;
          description += `**Carteira:** [\`${wallet.address}\`](https://polymarket.com/profile/${wallet.address})`;

          const tradeValue = trade.usdcSize || trade.price * trade.amount;
          const entryTier = getEntryTier(tradeValue);

          // Cria o embed
          const embed = new EmbedBuilder()
            .setTitle(`${emoji} ${typeLabel}`)
            .setURL(marketUrl)
            .setColor(color)
            .setDescription(description)
            .addFields(
              {
                name: `${entryTier.emoji} Entrada`,
                value: `${entryTier.label} ($${tradeValue.toFixed(2)})`,
                inline: true
              },
              {
                name: "💵 Preço",
                value: `$${trade.price.toFixed(3)}`,
                inline: true
              },
              {
                name: "📊 Shares",
                value: `${trade.amount.toFixed(1)}`,
                inline: true
              },
              {
                name: "💰 Valor",
                value: `$${tradeValue.toFixed(2)}`,
                inline: true
              }
            )
            .setFooter({ text: `Polymarket Tracker` })
            .setTimestamp(new Date(trade.timestamp * 1000));

          // P&L realizado (apenas em vendas)
          if (trade.realizedPnl !== undefined && trade.side === "SELL") {
            if (trade.realizedPnl > 0) {
              embed.addFields({ name: "🏆 Profit", value: `$${trade.realizedPnl.toFixed(2)}`, inline: true });
            } else if (trade.realizedPnl < 0) {
              embed.addFields({ name: "📉 Loss", value: `-$${Math.abs(trade.realizedPnl).toFixed(2)}`, inline: true });
            }
          }

          // Adiciona imagem do mercado se disponível
          if (trade.marketImageUrl && trade.marketImageUrl.length > 0) {
            embed.setImage(trade.marketImageUrl);
          }

          // Envia para todos os canais inscritos
          let sentCount = 0;
          let filteredCount = 0;
          let failedCount = 0;
          for (const sub of subs) {
            // --- Logica de filtros ---
            if (sub.filters) {
              const { minUsd, keywords } = sub.filters;

              if (minUsd && minUsd > 0 && tradeValue < minUsd) {
                filteredCount++;
                logger.debug(`   Canal ${sub.channelId} filtrado por valor ($${tradeValue.toFixed(2)} < $${minUsd})`);
                continue;
              }

              if (keywords && keywords.length > 0) {
                const titleLower = trade.marketTitle.toLowerCase();
                const hasKeyword = keywords.some((k) => titleLower.includes(k.toLowerCase()));
                if (!hasKeyword) {
                  filteredCount++;
                  logger.debug(`   Canal ${sub.channelId} filtrado por keyword: "${trade.marketTitle}"`);
                  continue;
                }
              }
            }

            try {
              const channel = await client.channels.fetch(sub.channelId).catch((error: any) => {
                logger.error(
                  `   Erro ao buscar canal ${sub.channelId}: ${error?.code ?? "UNKNOWN"} ${error?.message ?? error}`
                );
                return null;
              });

              if (!channel) {
                failedCount++;
                logger.error(`   Canal ${sub.channelId} nao encontrado ou inacessivel pelo bot`);
                continue;
              }

              if (!channel.isTextBased() || !("send" in channel) || typeof channel.send !== "function") {
                failedCount++;
                logger.error(`   Canal ${sub.channelId} nao aceita envio de mensagens`);
                continue;
              }

              const sendableChannel = channel as typeof channel & {
                send: (options: { embeds: EmbedBuilder[] }) => Promise<unknown>;
              };

              await sendableChannel.send({ embeds: [embed] });
              sentCount++;
              logger.info(`   Enviado para canal ${sub.channelId}`);
            } catch (e: any) {
              failedCount++;
              logger.error(`   Erro ao enviar para ${sub.channelId}: ${e?.code ?? "UNKNOWN"} ${e?.message ?? e}`);
            }
          }

          if (sentCount > 0) {
            // Marca como enviado
            sentMessages.set(trade.id, Date.now());
            logger.info(`   Mensagem enviada com sucesso para ${sentCount} canal(is)\n`);
          } else {
            logger.warn(
              `   Nenhum canal recebeu a mensagem (filtrados=${filteredCount}, falhas=${failedCount}, inscritos=${subs.length})\n`
            );
          }

          // Delay entre mensagens
          await new Promise((r) => setTimeout(r, 500));
        }

        // Pausa entre carteiras
        await new Promise((r) => setTimeout(r, 1000));
      }

      logger.debug(`✓ Ciclo de verificação concluído\n`);
    } catch (e) {
      logger.error("❌ Erro no Loop do Tracker:", e);
    }
  }, CHECK_INTERVAL);
}
